### Brief Overview 

* In this assignment, you will practice some data wrangling functionality commonly used in real-world projects.

* This assignment is designed around two dataset

* The Medicaid Data per State  

* IRS Statistics of Income (SOI) dataset


* The final product of this assignment will be a table with medication cost per Medicaid enrollee per state. This dataset will allow us to answer such questions as:
  * Which medications account for the bulk of a state's spending   
  * Which medications are prescribed more frequently in one state than another.
  * etc.

* The following URLs provide the datasets required for this assignment:
  * [medicaid_data.csv](https://www.dropbox.com/s/u6ctjqxnk0wxpbk/medicaid_data.csv?dl=1)
  * [medicaid_enrollment.tsv](https://www.dropbox.com/s/969mohzhpu78r10/medicaid_enrollment.tsv?dl=1)
  * [tax_data.csv](https://www.dropbox.com/s/zm37nu7vnirha4m/tax_data.csv?dl=1)


 ### Instructions on answering the questions

- Some of the questions below require that you only use methods or properties of a `DataFrame` or a `Series`. Therefore, any solution that uses a function that is not a method or property of `DataFrame` or `Series` will not be accepted, even when the solution yields an appropriate answer to the question.

- For instance, if you are asked to find the number of entries in the DataFrame `tax_data ` using only the DataFrame's methods or properties, then `len(tax_data)` is not an acceptable solution since `len()` is not a `tax_data` method. The statements below are both correct answers:

  - `info()` is a method `tax_data`
  
```python
   
    tax_data.info()
```

  or

  - `shape` is a property of `tax_data`


```python
tax_data.shape
```

- Similarly, if you are asked to find to count the number of unique entries in the `STATEFIPS` column of the `tax_data` DataFrame, then solutions using set() or len() are not acceptable for the following reasons:

- The solution below uses `set()` and `len()`, which are not `tax_data` methods

```python
len(set(tax_data["STATEFIPS"]))
```

- The solution below uses `unique()`, which is a  `tax_data` method, but then counts the number of uniques entries using `len()` which is not `tax_data` method

```python
len(tax_data["STATEFIPS"].unique())
```


- The statement below is an acceptable solution since it uses `nunique()` which is a method of the `Series` generated by indexing on a column of `tax_data` (`tax_data['STATEFIPS'])

```python
tax_data['STATEFIPS'].nunique()
```


- Chaining methods and properties is encouraged if it does not cause ambiguity 

- For instance, to identify whether a value is part of the index, write the following:

```python
tax_data.index.contains(99999)
```

- Rather than:

```python
9999 in tax_data.index
```

- You can only import `pandas` and `numpy`

- If you are not explicitly asked to only use methods or properties of a `DataFrame` or a `Series`, then any solution that does not rely on external modules will be accepted.

- __Important__: Provide the exact statement(s) used to answer each question

- Unless otherwise specified, each cell can contain multiple lines of code

* Finally, note that not all the functions necessary for answering the questions below were covered in class. As such, assuming you are using Jupyter Notebooks, I suggest you use `.SHIFT+TAB` on objects liberally to see which methods and properties are available on objects. If you are unsure what a method does, use `SHIFT+TAB` twice to invoke the `docstring`, or documentation for that function. This is not only a good way to see which functionality can be used to answer the questions below but also a great way to familiarize yourself with the plethora of functionality available through the `pandas` package. 


### 1. Loading and exploring the tax data

Import Pandas and numpy using their appropriate alias here


In [1]:
# WRITE YOUR CODE HERE

import pandas as pd
import numpy as np


* Load the IRS Statistics of Income (SOI) dataset (`tax_data.csv`) into a `DataFrame` called `tax_data`. This file can be downloaded from the `URL` provided above.

* This file `tax_data.csv` was preprocessed from the original file I obtained at the following URL:
  * [15zpallagi.csv](https://www.irs.gov/pub/irs-soi/15zpallagi.csv}

In [2]:
## WRITE YOUR CODE HERE 

tax_data = pd.read_csv('tax_data.csv')

* Use a `tax_data` method or property to display the first eight (8) rows of the `DataFrame`

In [3]:
## WRITE YOUR CODE HERE 

tax_data[:8]

STATEFIPS STATE  zipcode  agi_stub        N1     mars1     MARS2     MARS4  \
0        1.0    AL      0.0       1.0  836320.0  481570.0  109790.0  233260.0   
1        1.0    AL      0.0       2.0  494830.0  206630.0  146250.0  129390.0   
2        1.0    AL      0.0       3.0  261250.0   80720.0  139280.0   36130.0   
3        1.0    AL      0.0       4.0  166690.0   28510.0  124650.0   10630.0   
4        1.0    AL      0.0       5.0  212660.0   19520.0  184320.0    4830.0   
5        1.0    AL      0.0       6.0   55360.0    2950.0   49260.0     350.0   
6        1.0    AL  35004.0       1.0    1490.0     970.0     230.0     280.0   
7        1.0    AL  35004.0       2.0    1350.0     630.0     360.0     300.0   

       PREP         N2  ...    N10300     A10300   N85530   A85530   N85300  \
0  455560.0  1356760.0  ...  373410.0   328469.0      0.0      0.0      0.0   
1  275920.0  1010990.0  ...  395880.0   965011.0      0.0      0.0      0.0   
2  155100.0   583910.0  ...  251490.0  1333418.0      0.0      0.0      0.0   
3   99950.0   423990.0  ...  165320.0  1414283.0      0.0      0.0      0.0   
4  126860.0   589490.0  ...  212000.0  3820152.0    420.0    168.0     60.0   
5   41410.0   160530.0  ...   55300.0  6027793.0  22090.0  39519.0  27550.0   
6     700.0     2160.0  ...     690.0      610.0      0.0      0.0      0.0   
7     610.0     2540.0  ...    1140.0     3019.0      0.0      0.0      0.0   

    A85300   N11901    A11901    N11902     A11902  
0      0.0  61920.0   48150.0  732670.0  1933120.0  
1      0.0  73720.0  107304.0  415410.0  1187403.0  
2      0.0  64200.0  139598.0  193030.0   536699.0  
3      0.0  45460.0  128823.0  116440.0   377177.0  
4     31.0  83330.0  421004.0  121570.0   483682.0  
5  95112.0  28590.0  791573.0   15960.0   250289.0  
6      0.0    120.0      94.0    1290.0     2792.0  
7      0.0    210.0     301.0    1130.0     2935.0  

[8 rows x 131 columns]

*  Modify `tax_data` to uppercase all the header names. 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by uppercasing the columns yourself
  *  You can use `tax_data.columns`, which returns a `Series` of the column names.

* The resulting column name should look as follows:

```
STATEFIPS    STATE    ZIPCODE    AGI_STUB    N1    MARS1    MARS2    MARS4    PREP    N2    ...    
```

* Standardizing column names is a good way to avoid having to guess whether the column header is in lowercase or uppercase.

In [4]:
## WRITE YOUR CODE HERE

tax_data.columns = tax_data.columns.str.upper()
tax_data.columns

Index(['STATEFIPS', 'STATE', 'ZIPCODE', 'AGI_STUB', 'N1', 'MARS1', 'MARS2',
       'MARS4', 'PREP', 'N2',
       ...
       'N10300', 'A10300', 'N85530', 'A85530', 'N85300', 'A85300', 'N11901',
       'A11901', 'N11902', 'A11902'],
      dtype='object', length=131)

* What is the total number of entries (also called observations or instances) in `tax_data` dataset?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [5]:
## WRITE YOUR CODE HERE 

tax_data.shape[0]

166698

* If `STATEFIPS` is the header label of the first column in `tax_data`, write code to print the title of the 32nd column.  
  * You should use a single python expression
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [6]:
## WRITE YOUR CODE HERE 

tax_data.columns[31]

'N00900'

* If `STATEFIPS` is the first column label (index 0), what is the index of the column labeled `N10300`?
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [7]:
## WRITE YOUR CODE HERE 

tax_data.columns.get_loc('N10300')

121

- Print the count of unique zip codes in each state using descending order.
  - Your answer can only use `DataFrame` or `Series` methods or properties

- The result should look like the following ( '...' represents remaining data that is not shown here)
```python

        ZIPCODE
STATE	
TX	     9714
NY	     9238
CA	     8908
PA	     8208
IL	     7386
...
```

* The above indicates that there are 9,714 unique zip codes in TX, 9,238 in NY, etc.


In [31]:
## WRITE YOUR CODE HERE 

unique_zipcodes = tax_data.groupby('STATE')['ZIPCODE'].nunique().sort_values(ascending=False)
unique_zipcodes

STATE
TX    1619
NY    1540
CA    1485
PA    1368
IL    1231
OH     998
FL     920
MI     892
MO     890
IA     827
VA     793
MN     792
NC     725
WI     714
IN     674
GA     667
KY     656
KS     598
TN     590
AL     577
NJ     547
OK     547
WV     512
WA     497
AR     490
NE     486
MA     482
LA     453
MD     405
CO     395
SC     376
MS     370
ME     370
OR     356
AZ     295
ND     287
SD     287
CT     264
VT     238
NH     232
MT     225
ID     213
NM     209
UT     186
NV     128
WY     109
RI      71
HI      60
DE      57
AK      56
DC      24
Name: ZIPCODE, dtype: int64

* Identify the position of `HI` in the list of zip code counts per state generated in the previous question
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [32]:
## WRITE YOUR CODE HERE 

unique_zipcodes.index.get_loc('HI')

47

### 2. Identifying and Removing Ambiguous Zip Codes

- Count the number of entries where ZIPCODE is 0, assign your results to a variable named  `nb_invalid_zip`

In [10]:
## WRITE YOUR CODE HERE 

nb_invalid_zip = tax_data[tax_data.ZIPCODE == 0].shape[0]
nb_invalid_zip

306

* Run the line in the code cell below to make sure that `nb_invalid_zip` is an integer (`int`)
  * Note that `assert` will only print an error if `type(nb_invalid_zip)` is not of type `int`
* If the code cell below returns an error, then change your answer above so that the value returned is a number.

In [11]:
assert(type(nb_invalid_zip) == int)

* Remove from `tax_data` all the instances where the zip code is `0` and save the resulting `DataFrame` to a variable named `tax_data_valid_zip`

  * Your answer can only use `DataFrame` or `Series` methods or properties


In [55]:
## WRITE YOUR CODE HERE 

tax_data_valid_zip = tax_data[tax_data.ZIPCODE != 0]

* Run the line below to confirm that the operation worked as expected

  * The assertion below verifies that the number of lines with "zip code equal to  0" + number of lines in `tax_data_valid_zip` is equal to the number of lines in the original `DataFrame` `tax_data`
  
  * The assertion below will fail (and print an error message) if the results do not match. If that is the case, please review your code above.

In [88]:
assert((tax_data_valid_zip.shape[0] + nb_invalid_zip) == tax_data.shape[0])

### Identifying and Removing Lines with Missing Values

* How many lines contain at least one missing value `NaN` in the `tax_data_valid_zip` `DataFrame`?
  * Your answer can only use `DataFrame` methods and properties
* Assing the count of `NaN` into a variable called nb_missing_values

In [33]:
## WRITE YOUR CODE HERE 

nb_missing_values = tax_data_valid_zip.isnull().any(axis=1).sum()
nb_missing_values

199

* Create a new `DataFrame` containing all the lines from `tax_data_valid_zip`, except lines containing missing values (`NaN`)

* Call the new `DataFrame` `tax_data_valid_zip_cleaned

In [56]:
## WRITE YOUR CODE HERE 

tax_data_valid_zip_cleaned = tax_data_valid_zip[~tax_data_valid_zip.isnull().any(axis=1)]

* Run the line below to confirm that the operation worked as expected. The assertion below tests that:  
`nb_missing_values` + number of lines in `tax_data_valid_zip_cleaned` is equal to the number of lines in `tax_data_valid_zip`
  
* Note that assert will only print an error if the results do not match

In [35]:
assert((tax_data_valid_zip_cleaned.shape[0] + nb_missing_values) == tax_data_valid_zip.shape[0])

### Computing the Percentile Income per Zip Code
* The function `compute_percentile_zipcode` below computes the percentile income per zip code
* By default `percentile=0.5`, i.e., the function computes the median
* Read the code and make sure you understand what it does before moving on to the next question

In [49]:
def compute_percentile_zip(df_zip, percentile=0.5):     
    index_median = sum(( df_zip["N1"]/ sum(df_zip["N1"])).cumsum() <= percentile)
    val_below_or_at_median = (df_zip["A00100"] /df_zip["N1"]).iloc[index_median]
    return val_below_or_at_median

* Compute the 65th income percentile ( percentile=0.65) for each zip code in `tax_data_valid_zip_cleaned` data frame

  * Your answer can only use `DataFrame` or `Series` methods or properties
  
  * You can use any of the submethods discussed in the Split-Apply-Combine paradigm.
  
* Sort the values in descending order and assign them to a new `DataFrame` called `zip_rev_all`

* The resulting `Series` should look like the following ( '...' represents remaining data that is not shown)

```Python
ZIPCODE
33109    3954.114286
33480    3413.301538
...
```

In [48]:
## WRITE YOUR CODE HERE



* What are the three zip codes with the most significant 65th percentile value for income?
 * I.e., the top three zip codes in your sorted listed

In [ ]:
## WRITE YOUR CODE HERE 

# 2 Working with the Medicaid Data

### Loading and exploring the data 

* Load the Medicaid data stored in the file `medicaid_data.csv` into a `DataFrame` called `medicaid_data`. The file can be downloaded from the URL provided above. 
* Note that this is quite large and may take some time to load on a computer with modest RAM resources (e.g., less than 6-8 GB)


In [50]:
## WRITE YOUR CODE HERE

medicaid_data = pd.read_csv('medicaid_data.csv')
medicaid_data[:5]

Utilization Type State          NDC Product Name  Units Reimbursed  \
0             MCOU    PA  55150023930   Dexamethas              33.0   
1             FFSU    NY     23917710   ALPHAGAN P             570.0   
2             MCOU    OR  13925050501   Dapsone 10             456.0   
3             FFSU    MN  51862006401     DIAZEPAM             780.0   
4             FFSU    MN    781237101   DEXTROAMPH             451.0   

   Number of Prescriptions  Total Amount Reimbursed  \
0                     19.0                   234.98   
1                     57.0                 16006.34   
2                     15.0                  1052.42   
3                     16.0                    89.60   
4                     12.0                  1411.24   

   Medicaid Amount Reimbursed  Non Medicaid Amount Reimbursed  \
0                      234.98                            0.00   
1                    16006.34                            0.00   
2                     1052.42                            0.00   
3                       77.60                           12.00   
4                      198.93                         1212.31   

               Location  
0    (40.5773, -77.264)  
1   (42.1497, -74.9384)  
2  (44.5672, -122.1269)  
3   (45.7326, -93.9196)  
4   (45.7326, -93.9196)

* Modify `medicaid_data` to capitalize all column names 

  * If your solution uses an assignment, the righthand side of the assignment (rvalue) can only use `DataFrame` or `Series` methods or properties


In [51]:
## WRITE YOUR CODE HERE

medicaid_data.columns = medicaid_data.columns.str.upper()
medicaid_data.columns

Index(['UTILIZATION TYPE', 'STATE', 'NDC', 'PRODUCT NAME', 'UNITS REIMBURSED',
       'NUMBER OF PRESCRIPTIONS', 'TOTAL AMOUNT REIMBURSED',
       'MEDICAID AMOUNT REIMBURSED', 'NON MEDICAID AMOUNT REIMBURSED',
       'LOCATION'],
      dtype='object')

* Familiarize yourself with the data
  * the `NDC` column stands for National Drug Code, a universal product identifier for human drugs in the United States
  * The remaining column names are self-explanatory
* Explore the number of lines and columns in the data
* Check that your column headers are capitalized

In [52]:
## WRITE YOUR CODE HERE

medicaid_data.shape

(1695546, 10)

* If you explore the `LOCATION` column for all the entries for which "STATE" value is equal to "HI" you'll notice that all the values are identical.
* Are there any states that have more than one value for `LOCATION`. 
  * This question can be answered by using aggregation and the Split-Apply-Combine paradigm
  * Your answer can only use `DataFrame` or `Series` methods or properties

In [57]:
## WRITE YOUR CODE HERE 

medicaid_data.groupby('STATE')['LOCATION'].nunique()

STATE
AK    1
AL    1
AR    1
AZ    1
CA    1
CO    1
CT    1
DC    1
DE    1
FL    1
GA    1
HI    1
IA    1
ID    1
IL    1
IN    1
KS    1
KY    1
LA    1
MA    1
MD    1
ME    1
MI    1
MN    1
MO    1
MS    1
MT    1
NC    1
ND    1
NE    1
NH    1
NJ    1
NM    1
NV    1
NY    1
OH    1
OK    1
OR    1
PA    1
RI    1
SC    1
SD    1
TN    1
TX    1
UT    1
VA    1
VT    1
WA    1
WI    1
WV    1
WY    1
XX    0
Name: LOCATION, dtype: int64

There are no states that have more than one value for LOCATION.

* To compare medication prescriptions across states in a fair and balanced way, we need the number of Medicaid beneficiaries in each state.  That is, we need to compare per capita. The following example illustrates the importance of normalizing the values `UNITS REIMBURSED` for each medication in each state by the number of Medicaid enrollees in each state.

 
* The `medicaid_data` DataFrame shows that for the drug with NDC `61958180101` (the drug name is HARVONI and it's used to treat Hepatitis C) there were 11,886  units sold in KY, versus 40,142 in CA -- that's almost 4 times more units sold in CA compared to KY. However, there are 1,284,193 Medicaid enrollees in KY, versus 13,096,861 in California. If we normalize the number of units sold in KY, versus CA, we find that the normalized there were close to 3 times more HARVONI prescriptions in KY  than in CA. This is ___perhaps___ justified by the fact the KY has one of the highest rates of reported cases of Hepatitis C in the US (2.7% in KY versus 0.2% in CA).


https://www.cdc.gov/hepatitis/statistics/2015surveillance/pdfs/2015hepsurveillancerpt.pdf

* The number of enrollees per state was obtained [here](https://www.medicaid.gov/medicaid/managed-care/enrollment/index.html)

* A parsed/processed version (`medicaid_enrollment.tsv`) can be downloaded from the URL provided above. Use `pandas` to load the `medicaid_enrollment.tsv` file into DataFrame named `medicaid_enrollment`

In [58]:
## WRITE YOUR CODE HERE

medicaid_enrollment = pd.read_csv('medicaid_enrollment.tsv', sep='\t')
medicaid_enrollment

STATE Total Medicaid Enrollees
0                    Alabama                1,050,989
1                     Alaska                  164,783
2             American Samoa                      NaN
3                    Arizona                1,740,520
4                   Arkansas                  762,166
5                 California               13,096,861
6                   Colorado                1,264,600
7                Connecticut                  746,119
8                   Delaware                  227,909
9       District of Columbia                  271,428
10                   Florida                3,808,334
11                   Georgia                1,990,810
12                      Guam                      n/a
13                    Hawaii                  340,513
14                     Idaho                  283,355
15                  Illinois                3,269,999
16                   Indiana                1,295,358
17                      Iowa                  618,505
18                    Kansas                  403,844
19                  Kentucky                1,284,193
20                 Louisiana                1,402,212
21                     Maine                  288,324
22                  Maryland                1,271,445
23             Massachusetts                1,829,618
24                  Michigan                3,947,031
25                 Minnesota                1,052,521
26               Mississippi                  740,937
27                  Missouri                  944,257
28                   Montana                  139,950
29                  Nebraska                  239,463
30                    Nevada                  588,304
31             New Hampshire                  186,399
32                New Jersey                1,705,594
33                New Mexico                  826,155
34                  New York                6,281,038
35            North Carolina                1,965,805
36              North Dakota                   86,250
37  Northern Mariana Islands                      NaN
38                      Ohio                3,060,446
39                  Oklahoma                  829,561
40                    Oregon                1,123,913
41              Pennsylvania                2,569,232
42               Puerto Rico                1,458,819
43              Rhode Island                  308,521
44            South Carolina                1,233,430
45              South Dakota                  124,497
46                 Tennessee                1,562,745
47                     Texas                4,273,982
48                      Utah                  293,867
49                   Vermont                  206,469
50            Virgin Islands                      NaN
51                  Virginia                1,092,225
52                Washington                1,771,679
53             West Virginia                  545,748
54                 Wisconsin                1,209,714
55                   Wyoming                   66,532

* Modify `medicaid_enrollment` `DataFrame` to capitalize all the column names 
  * Your answer can only use `DataFrame` or `Series` methods or properties
  * Do not hardcode the operation by manually uppercasing the columns

In [59]:
## WRITE YOUR CODE HERE

medicaid_enrollment.columns = medicaid_enrollment.columns.str.upper()
medicaid_enrollment.columns

Index(['STATE', 'TOTAL MEDICAID ENROLLEES'], dtype='object')

* Note that some states/territories have missing values. Remove the missing values and save the resulting `DataFrame` as a new variable named `medicaid_enrollment_cleaned`
* Pay attention to how 'n/a' is given here!
* After cleaning, do you still have the Guam entry? If so, reconsider what missing value means in this context

In [83]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned = medicaid_enrollment[~medicaid_enrollment['TOTAL MEDICAID ENROLLEES'].str.contains('n/a', na=False)]
medicaid_enrollment_cleaned = medicaid_enrollment_cleaned.dropna()
medicaid_enrollment_cleaned

STATE TOTAL MEDICAID ENROLLEES
0                Alabama                1,050,989
1                 Alaska                  164,783
3                Arizona                1,740,520
4               Arkansas                  762,166
5             California               13,096,861
6               Colorado                1,264,600
7            Connecticut                  746,119
8               Delaware                  227,909
9   District of Columbia                  271,428
10               Florida                3,808,334
11               Georgia                1,990,810
13                Hawaii                  340,513
14                 Idaho                  283,355
15              Illinois                3,269,999
16               Indiana                1,295,358
17                  Iowa                  618,505
18                Kansas                  403,844
19              Kentucky                1,284,193
20             Louisiana                1,402,212
21                 Maine                  288,324
22              Maryland                1,271,445
23         Massachusetts                1,829,618
24              Michigan                3,947,031
25             Minnesota                1,052,521
26           Mississippi                  740,937
27              Missouri                  944,257
28               Montana                  139,950
29              Nebraska                  239,463
30                Nevada                  588,304
31         New Hampshire                  186,399
32            New Jersey                1,705,594
33            New Mexico                  826,155
34              New York                6,281,038
35        North Carolina                1,965,805
36          North Dakota                   86,250
38                  Ohio                3,060,446
39              Oklahoma                  829,561
40                Oregon                1,123,913
41          Pennsylvania                2,569,232
42           Puerto Rico                1,458,819
43          Rhode Island                  308,521
44        South Carolina                1,233,430
45          South Dakota                  124,497
46             Tennessee                1,562,745
47                 Texas                4,273,982
48                  Utah                  293,867
49               Vermont                  206,469
51              Virginia                1,092,225
52            Washington                1,771,679
53         West Virginia                  545,748
54             Wisconsin                1,209,714
55               Wyoming                   66,532

### Converting `TOTAL MEDICAID ENROLLEE` Data Type

* Given that data on `TOTAL MEDICAID ENROLLEE` column contains commas on file (ex. 3,269,999 instead of 3269999), `pandas` has erroneously set the data type for that column as a string. We need to convert the column from string to `int` since we will be using it in an arithmetic expression during normalization



In [85]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'] = medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].str.replace(',', '').astype(int)

* Inspect the `dtype` property of "TOTAL MEDICAID ENROLLEES" column, and make sure that the data type is `int`

In [86]:
## WRITE YOUR CODE HERE

medicaid_enrollment_cleaned['TOTAL MEDICAID ENROLLEES'].dtype

dtype('int64')

### Associating `medicaid_data` and `medicaid_enrollment_cleaned`

* We can use the shared State information across both tables to associate both tables (SQL JOIN).

* However, `medicaid_data` contains two-letter state abbreviations, while `medicaid_enrollment_cleaned` contains the complete state name

  * We need to convert (or append) two-letter state abbreviations to `medicaid_enrollment_cleaned`

* Pandas can read HTML and parse the code for tables. We will use that functionality to read in the state abbreviations from a Wikipedia page.

  * A brief description of what the code does is included in the comments
  
* Note that the code below may throw an error if `lxml` is not installed on your machine. You should install any Python packages that Python complains about before proceeding.

In [87]:

import requests

url = 'https://www.50states.com/abbreviations.htm'
header = {
  "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.75 Safari/537.36",
  "X-Requested-With": "XMLHttpRequest"
}

r = requests.get(url, headers=header)

tables = pd.read_html(r.text)


# We access the desired table by giving it's index.
# Since the URL contain only one table, then we can access that table using index 0
Codes_abbreviations = tables[0]
Codes_abbreviations.head(5)

US STATE POSTAL ABBREVIATION STANDARD ABBREVIATION
0     Alabama                  AL                  Ala.
1      Alaska                  AK                Alaska
2     Arizona                  AZ                 Ariz.
3    Arkansas                  AR                  Ark.
4  California                  CA                Calif.

* Change the the `DataFrame`'s headers from ['US STATE',	'POSTAL ABBREVIATION', 	'STANDARD ABBREVIATION'] to ['US STATE', 'ABBREVIATION', `STD ABBREVIATION`]
  * You can hard code this operation

In [99]:
## WRITE YOUR CODE HERE 

Codes_abbreviations = Codes_abbreviations.rename(columns={'POSTAL ABBREVIATION': 'ABBREVIATION', 'STANDARD ABBREVIATION': 'STD ABBREVIATION'})
Codes_abbreviations

US STATE ABBREVIATION STD ABBREVIATION
0          Alabama           AL             Ala.
1           Alaska           AK           Alaska
2          Arizona           AZ            Ariz.
3         Arkansas           AR             Ark.
4       California           CA           Calif.
5         Colorado           CO           Color.
6      Connecticut           CT            Conn.
7         Delaware           DE             Del.
8          Florida           FL             Fla.
9          Georgia           GA              Ga.
10          Hawaii           HI           Hawaii
11           Idaho           ID            Idaho
12        Illinois           IL             Ill.
13         Indiana           IN             Ind.
14            Iowa           IA             Iowa
15          Kansas           KS             Kan.
16        Kentucky           KY              Ky.
17       Louisiana           LA              La.
18           Maine           ME            Maine
19        Maryland           MD              Md.
20   Massachusetts           MA            Mass.
21        Michigan           MI            Mich.
22       Minnesota           MN            Minn.
23     Mississippi           MS            Miss.
24        Missouri           MO              Mo.
25         Montana           MT            Mont.
26        Nebraska           NE             Neb.
27          Nevada           NV             Nev.
28   New Hampshire           NH             N.H.
29      New Jersey           NJ             N.J.
30      New Mexico           NM             N.M.
31        New York           NY             N.Y.
32  North Carolina           NC             N.C.
33    North Dakota           ND             N.D.
34            Ohio           OH             Ohio
35        Oklahoma           OK            Okla.
36          Oregon           OR             Ore.
37    Pennsylvania           PA              Pa.
38    Rhode Island           RI             R.I.
39  South Carolina           SC             S.C.
40    South Dakota           SD           S.Dak.
41       Tennessee           TN            Tenn.
42           Texas           TX             Tex.
43            Utah           UT             Utah
44         Vermont           VT             V.T.
45        Virginia           VA              Va.
46      Washington           WA            Wash.
47   West Virginia           WV            W.Va.
48       Wisconsin           WI             Wis.
49         Wyoming           WY             Wyo.

* Combine the tables `medicaid_enrollment_cleaned` and `Codes_abbreviations` such that the resulting `DataFrame` contains all the columns in `medicaid_enrollment_cleaned` and only `ABBREVIATION` from `Codes_abbreviations` 

* Save the results to a variable named `medicaid_enrollment_cleaned_with_zip`

* `medicaid_enrollment_cleaned_with_zip` should look like the following ( '...' represents remaining data that is not shown):


```
   STATE    Total Medicaid Enrollees    ABBREVIATION
0    Alabama    1,050,989    AL
1    Alaska    164,783    AK
2    Arizona    1,740,520    AZ
3    Arkansas    762,166    AR
4    California    13,096,861    CA
...
```

* We did not cover joins in class -- you find a plethora of examples on how to do this online (e.g., `https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html`) and in a notebook under `Data Wrangling - Advanced` module. See for instance:


* If you cannot get it to work, contact the `TA` for the solution. You will not be penalized if you don't answer this question. 

In [116]:
## WRITE YOUR CODE HERE 

medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned.merge(Codes_abbreviations, how='inner', left_on='STATE', right_on='US STATE')
medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop(medicaid_enrollment_cleaned_with_zip.columns[[2, 4]], axis = 1)
medicaid_enrollment_cleaned_with_zip

STATE  TOTAL MEDICAID ENROLLEES ABBREVIATION
0          Alabama                   1050989           AL
1           Alaska                    164783           AK
2          Arizona                   1740520           AZ
3         Arkansas                    762166           AR
4       California                  13096861           CA
5         Colorado                   1264600           CO
6      Connecticut                    746119           CT
7         Delaware                    227909           DE
8          Florida                   3808334           FL
9          Georgia                   1990810           GA
10          Hawaii                    340513           HI
11           Idaho                    283355           ID
12        Illinois                   3269999           IL
13         Indiana                   1295358           IN
14            Iowa                    618505           IA
15          Kansas                    403844           KS
16        Kentucky                   1284193           KY
17       Louisiana                   1402212           LA
18           Maine                    288324           ME
19        Maryland                   1271445           MD
20   Massachusetts                   1829618           MA
21        Michigan                   3947031           MI
22       Minnesota                   1052521           MN
23     Mississippi                    740937           MS
24        Missouri                    944257           MO
25         Montana                    139950           MT
26        Nebraska                    239463           NE
27          Nevada                    588304           NV
28   New Hampshire                    186399           NH
29      New Jersey                   1705594           NJ
30      New Mexico                    826155           NM
31        New York                   6281038           NY
32  North Carolina                   1965805           NC
33    North Dakota                     86250           ND
34            Ohio                   3060446           OH
35        Oklahoma                    829561           OK
36          Oregon                   1123913           OR
37    Pennsylvania                   2569232           PA
38    Rhode Island                    308521           RI
39  South Carolina                   1233430           SC
40    South Dakota                    124497           SD
41       Tennessee                   1562745           TN
42           Texas                   4273982           TX
43            Utah                    293867           UT
44         Vermont                    206469           VT
45        Virginia                   1092225           VA
46      Washington                   1771679           WA
47   West Virginia                    545748           WV
48       Wisconsin                   1209714           WI
49         Wyoming                     66532           WY

* We have no further use for the column STATE in  `medicaid_enrollment_cleaned_with_zip`

  * Remove the column and make sure the data in `medicaid_enrollment_cleaned_with_zip` looks like the following  ( `...` represents remaining data that is not shown):



```
    Total Medicaid Enrollees    ABBREVIATION
0   1,050,989                  AL
1   164,783                    AK
2   1,740,520                  AZ
3   762,166                    AR
4   13,096,861                 CA
....

```

In [118]:
## WRITE YOUR CODE HERE 

medicaid_enrollment_cleaned_with_zip = medicaid_enrollment_cleaned_with_zip.drop(['STATE'], axis = 1)
medicaid_enrollment_cleaned_with_zip

TOTAL MEDICAID ENROLLEES ABBREVIATION
0                    1050989           AL
1                     164783           AK
2                    1740520           AZ
3                     762166           AR
4                   13096861           CA
5                    1264600           CO
6                     746119           CT
7                     227909           DE
8                    3808334           FL
9                    1990810           GA
10                    340513           HI
11                    283355           ID
12                   3269999           IL
13                   1295358           IN
14                    618505           IA
15                    403844           KS
16                   1284193           KY
17                   1402212           LA
18                    288324           ME
19                   1271445           MD
20                   1829618           MA
21                   3947031           MI
22                   1052521           MN
23                    740937           MS
24                    944257           MO
25                    139950           MT
26                    239463           NE
27                    588304           NV
28                    186399           NH
29                   1705594           NJ
30                    826155           NM
31                   6281038           NY
32                   1965805           NC
33                     86250           ND
34                   3060446           OH
35                    829561           OK
36                   1123913           OR
37                   2569232           PA
38                    308521           RI
39                   1233430           SC
40                    124497           SD
41                   1562745           TN
42                   4273982           TX
43                    293867           UT
44                    206469           VT
45                   1092225           VA
46                   1771679           WA
47                    545748           WV
48                   1209714           WI
49                     66532           WY

- Use `DataFrame medicaid_enrollment_cleaned_with_zip` to assign the appropriate number of Medicaid enrollees to each entry in the `medicaid_data`

   I.E., instead of the 10 original columns, `medicaid_data` will now have an 11th column representing the `TOTAL MEDICAID ENROLLEES` according to the STATE value in the entry.
   
- Save the resulting DataFrame into a new variable called `medicaid_data_w_enrollments`
- The resulting DataFrame should look like the following (`...` represents remaining data that is not shown):

```
UTILIZATION TYPE    STATE    NDC    PRODUCT NAME    UNITS REIMBURSED    NUMBER OF PRESCRIPTIONS    TOTAL AMOUNT REIMBURSED    MEDICAID AMOUNT REIMBURSED    NON MEDICAID AMOUNT REIMBURSED    LOCATION
0    MCOU    PA    55150023930    Dexamethas    33.0    19.0    234.98    234.98    0.0    (40.5773, -77.264)
1    FFSU    NY    23917710    ALPHAGAN P    570.0    57.0    16006.34    16006.34    0.0    (42.1497, -74.9384)
2    MCOU    OR    13925050501    Dapsone 10    456.0    15.0    1052.42    1052.42    0.0    (44.5672, -122.1269)
...
```

* The order of the columns in the `DataFrame` is not relevant. This answer uses the same approach as the one used to `merge` the tables above. 

In [ ]:
## WRITE YOUR CODE HERE 


- Remove any lines where "STATE" or "PRODUCT NAME" are missing from  `medicaid_data_w_enrollments`

In [ ]:
## WRITE YOUR CODE HERE 



- Use ["STATE", "PRODUCT NAME"] as hierarchical index for `medicaid_data_w_enrollments`. Recall that a hierarchical index is simply an index with multiple levels of indexing (multiple columns)
  * Hint: the function to set an index on a `DataFrame` can take a single column name or a list of column names. The list here is  ["STATE", "NDC"]
- Call the new data `medicaid_data_w_enrollments_hierarch`
- Inspect your data to make sure the new index has now two levels (STATE and NDC)

In [ ]:
## WRITE YOUR CODE HERE 



* Write a single Pandas expression to print all the lines containing `NDC` `61958180101` in the Pennsylvania ("PA")

 * Use a single indexing call (bracket notation) using `loc`
 
 * Hint 1: Since your index is hierarchical, `loc` is expecting two values, the first for STATE and the second for NDC


In [ ]:
## WRITE YOUR CODE HERE 


#### Computing the normalized `UNITS REIMBURSED` per drug 



* Compute the `UNITS REIMBURSED` for each "NDC" in each state normalized by the number of enrollees.



- For instance in `PA`, the total `UNITS REIMBURSED` for the HARVONI `NDC` (61958180101) is 10,612



```python

total_amount_reimbursed = medicaid_data_w_enrollments_hierarch.loc[("PA", 61958180101), "UNITS REIMBURSED"].sum() 

```



- And the numebr of Medicaid Enrollees in "PA" is 2,569,232





```python

total_enrollees_PA = medicaid_data_w_enrollments_hierarch.loc["PA", "TOTAL MEDICAID ENROLLEES"].unique()[0]

```

- Therefore, the UNITS REIMBURSED per enrollee for "HARVONI" is  0.00413041718303





```python

print(total_amount_reimbursed/total_enrollees_PA)

```



- Rather than work directly with the ratios, we are going to compute and work with their logarithm (np.log2) instead.



  - The reason we use logs here is to avoid working small numbers. More on log-transformation in future lectures



- Save the result in a `Series` called `medicaid_reimbursement_per_enrollee`





- Your final result should be a `Series` that looks like the following ( `...` represents data that is not shown here):
  * Note that negative values are the result of the `log` transformation 

```
STATE  NDC    
AK     2143380    -9.609109
       2143480   -10.008280
       2322730    -6.109830
       2322830    -4.444321
       2322930    -3.855995
...
AL     2143380    -9.940595
       2143480    -9.805485
       2322730    -5.336260
...
MA     2143380    -7.921997
       2143480    -7.803463
       2144511   -13.472194
       2197590    -7.741402
       2322730    -5.414724
       2322830    -4.592154
       2322930    -4.205626
...
```



In [ ]:
## WRITE YOUR CODE HERE 



- To facilitate working with the final data, we are going to unstack `medicaid_reimbursement_per_enrollee` into a variable called  `medicaid_norm_ndc`
- Using `medicaid_reimbursement_per_enrollee`, generate a `DataFrame` where: 
  - `index` is the two-letter state symbol 
  - column names are the `NDC` codes 
- The `DataFrame` should be formatted as in the image below
  - Hint: simply unstack the data


<img src="media/unstacked.png" alt="drawing" style="width:900px;"/>


In [ ]:
## WRITE YOUR CODE HERE 


#### Exploring the Data 

* What is the drug with the highest log-normalized ratio in Hawaii?


In [ ]:
## WRITE YOUR CODE HERE 


* Investigate the NDC of the product with the highest log-normalized ratio in Hawaii 
  * What is it used for?

* Compare the value of `Units Reimbursed` that product between HI and other states, (take for instance MA, FL, OR and WA)

In [ ]:
## WRITE YOUR CODE HERE 

* Can you think for reasons why this product has the highest log-normalized ratio in Hawaii? (Not graded)

In [ ]:
## WRITE YOUR ANSWER HERE 

* Find and list all unique `NDC`s for which the difference between the largest and second large log-scale ratio-by-state is at least 10.

* For instance:
  * The highest log-normalized ratio for `00591289749` (`AZACITIDIN`) is OK where it has a log-normalized ratio of `-1.025642`.
  * The second highest log-normalized ratio for `00591289749` is in `GA` where is has a log-normalized ratio of  `-12.623428`

In [ ]:
## WRITE YOUR CODE HERE 


* The Drug `AZACITIDINE` has a very high normalized UNITS REIMBURSED in OK compared to other states
   * Normalized log value is -1.025642 (or a ratio 0.49119167009735065)
   * The second highest state has a log value of -12.623428 (0.000158478197834722)
* Oklahoma is not a high-incidence state for cancer
* Could the following explain what is happening in Oklahoma?
  * The following is clinical trial conducted by the University of Oklahoma

https://www.centerwatch.com/clinical-trials/listings/92093/acute-myeloid-leukemia-aml-study-asp2215-gilteritinib-by/?&geo_lat=35.4675602&geo_lng=-97.5164276&radius=10